runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillEndpoints.java

```diff
@@ -0,0 +1,112 @@
+
+import com.google.auto.value.AutoValue;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableList;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+
+/** Value class for holding endpoints used for communicating with Windmill service. */
+@AutoValue
+public abstract class WindmillEndpoints {
+  public static WindmillEndpoints from(
+      Windmill.WorkerMetadataResponse workerMetadataResponseProto) {
+    ImmutableMap<String, WindmillEndpoints.Endpoint> globalDataServers =
+        workerMetadataResponseProto.getGlobalDataEndpointsMap().entrySet().stream()
+            .collect(
+                toImmutableMap(
+                    Map.Entry::getKey, // global data key
+                    endpoint -> WindmillEndpoints.Endpoint.from(endpoint.getValue())));
+
+    ImmutableList<WindmillEndpoints.Endpoint> windmillServers =
+        workerMetadataResponseProto.getWorkEndpointsList().stream()
+            .map(WindmillEndpoints.Endpoint::from)
+            .collect(toImmutableList());
+
+    return WindmillEndpoints.builder()
+        .setGlobalDataServers(globalDataServers)
+        .setWindmillServers(windmillServers)
+        .build();
+  }
+
+  public static WindmillEndpoints.Builder builder() {
+    return new AutoValue_WindmillEndpoints.Builder();
+  }
+
+  /**
+   * Used by GetData GlobalDataRequest(s) to support Beam side inputs. Returns a map where the key
+   * is a global data tag and the value is the endpoint where the data associated with the global
+   * data tag resides.
+   *
+   * @see <a href="https://beam.apache.org/documentation/programming-guide/#side-inputs">Beam Side
+   *     Inputs</a>
+   */
+  public abstract ImmutableMap<String, Endpoint> globalDataServers();
+
+  /**
+   * Used by GetWork/GetData/CommitWork calls to send, receive, and commit work directly to/from
+   * Windmill servers. Returns a list of endpoints used to communicate with the corresponding
+   * Windmill servers.
+   */
+  public abstract ImmutableList<Endpoint> windmillServers();
+
+  @AutoValue
+  public abstract static class Endpoint {
+    public static Endpoint.Builder builder() {
+      return new AutoValue_WindmillEndpoints_Endpoint.Builder();
+    }
+
+    public static Endpoint from(Windmill.WorkerMetadataResponse.Endpoint endpointProto) {
+      Endpoint.Builder endpointBuilder = Endpoint.builder();
+      if (endpointProto.hasDirectEndpoint() && !endpointProto.getDirectEndpoint().isEmpty()) {
+        endpointBuilder.setDirectEndpoint(endpointProto.getDirectEndpoint());
+      }
+      if (endpointProto.hasWorkerToken() && !endpointProto.getWorkerToken().isEmpty()) {
+        endpointBuilder.setWorkerToken(endpointProto.getWorkerToken());
+      }
+      return endpointBuilder.build();
+    }
+
+    public abstract Optional<String> directEndpoint();
```

Comment:
is there a better way to represent ipv6 address than String?

Reply:
lets use Inet6Address, i will update

Reply:
added